### PR TITLE
ENH: RangeIndex._shallow_copy can return RangeIndex

### LINKED ISSUE #43885: PERF: `Index.union` is materializing even when it is not needed when `sort=False`
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this issue exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this issue exists on the master branch of pandas.

### Reproducible Example

We are currently materializing `RangeIndex` which is already monotonically increasing when `sort=False`.

```python
>>> import pandas as pd
>>> pd.__version__
'1.3.3'
>>> idx1 = pd.RangeIndex(0, 10)
>>> idx2 = pd.RangeIndex(0, 5)
>>> idx1.union(idx2)
RangeIndex(start=0, stop=10, step=1)
>>> idx1.union(idx2, sort=False)
Int64Index([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], dtype='int64')
```

This behavior was introduced in this PR: https://github.com/pandas-dev/pandas/pull/25788

From the PR description, it seems to be a conscious decision, but for performance reasons we should ideally be checking the result of the `union` if it is monotonically increasing/decreasing and then decide to sort or not based on `sort` parameter. This alleviates materialization and thus memory pressure too.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 73c68257545b5f8530b7044f56647bd2db92e2ba
python           : 3.9.6.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 20.6.0
Version          : Darwin Kernel Version 20.6.0: Mon Aug 30 06:12:21 PDT 2021; root:xnu-7195.141.6~3/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : en_US.UTF-8
pandas           : 1.3.3
numpy            : 1.21.2
pytz             : 2021.1
dateutil         : 2.8.2
pip              : 21.2.4
setuptools       : 57.4.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 5.0.0
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### Prior Performance

_No response_

### LINKED ISSUE #46675: Index is being materialized in `pd.concat` when `axis=1`
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this issue exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this issue exists on the main branch of pandas.

### Reproducible Example

```python
>>> import pandas as pd
>>> s1 = pd.Series(["a", "b", "c"])
>>> s2 = pd.Series(["a", "b"])
>>> s3 = pd.Series(["a", "b", "c", "d"])
>>> s4 = pd.Series([])
<stdin>:1: FutureWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
>>> 
>>> sort = False
>>> join = 'outer'
>>> ignore_index = False
>>> axis = 1
>>> 
>>> result = pd.concat(
...         [s1, s2, s3, s4],
...         sort=sort,
...         join=join,
...         ignore_index=ignore_index,
...         axis=axis,
...     )
>>> 
>>> print(result)
     0    1  2   3
0    a    a  a NaN
1    b    b  b NaN
2    c  NaN  c NaN
3  NaN  NaN  d NaN
>>> print(result.index)
Int64Index([0, 1, 2, 3], dtype='int64')
>>> print(pd.__version__)
1.4.2
```

### Installed Versions

<details>

>>> pd.show_versions()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.8/site-packages/pandas/util/_print_versions.py", line 109, in show_versions
    deps = _get_dependency_info()
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.8/site-packages/pandas/util/_print_versions.py", line 88, in _get_dependency_info
    mod = import_optional_dependency(modname, errors="ignore")
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.8/site-packages/pandas/compat/_optional.py", line 138, in import_optional_dependency
    module = importlib.import_module(name)
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 843, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.8/site-packages/setuptools/__init__.py", line 8, in <module>
    import _distutils_hack.override  # noqa: F401
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.8/site-packages/_distutils_hack/override.py", line 1, in <module>
    __import__('_distutils_hack').do_override()
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.8/site-packages/_distutils_hack/__init__.py", line 72, in do_override
    ensure_local_distutils()
  File "/nvme/0/pgali/envs/cudfdev/lib/python3.8/site-packages/_distutils_hack/__init__.py", line 59, in ensure_local_distutils
    assert '_distutils' in core.__file__, core.__file__
AssertionError: /nvme/0/pgali/envs/cudfdev/lib/python3.8/distutils/core.py
>>> 

</details>

### Prior Performance

```python
>>> import pandas as pd
... s1 = pd.Series(["a", "b", "c"])
... s2 = pd.Series(["a", "b"])
... s3 = pd.Series(["a", "b", "c", "d"])
... s4 = pd.Series([])
... 
... sort = False
... join = 'outer'
... ignore_index = False
... axis = 1
... 
... result = pd.concat(
...         [s1, s2, s3, s4],
...         sort=sort,
...         join=join,
...         ignore_index=ignore_index,
...         axis=axis,
...     )
... 
... print(result)
... print(result.index)
... print(pd.__version__)
... 
<input>:5: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
     0    1  2   3
0    a    a  a NaN
1    b    b  b NaN
2    c  NaN  c NaN
3  NaN  NaN  d NaN
RangeIndex(start=0, stop=4, step=1)
1.3.5
```

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #46675 (Replace xxxx with the Github issue number)
- [x] closes #43885 
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/v1.5.0.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/indexes/range.py [hunk 1]
```diff
@@ -435,6 +435,10 @@ def _shallow_copy(self, values, name: Hashable = no_default):
 
         if values.dtype.kind == "f":
             return Float64Index(values, name=name)
+        unique_diffs = np.unique(np.diff(values))
```
Comment (jbrockmendel) [inline pandas/core/indexes/range.py:438]: libalgos.unique_deltas?

Comment (mroeschke) [reply pandas/core/indexes/range.py:438]: Ah cool I'll try this out. 

This is more of an experiment to see what breaks when returning a RangeIndex. Apparently a lot of things

File: pandas/core/indexes/range.py [hunk 2]
```diff
@@ -435,7 +436,13 @@ def _shallow_copy(self, values, name: Hashable = no_default):
 
         if values.dtype.kind == "f":
             return Float64Index(values, name=name)
-        return Int64Index._simple_new(values, name=name)
+        unique_diffs = unique_deltas(values)
```
Comment (jreback) [inline pandas/core/indexes/range.py:439]: can you add a comment here on why you are doing this

**Review by jreback — CHANGES_REQUESTED** [review]
asv's prove out the change in perf?

**Comment by mroeschke** [comment]
> asv's prove out the change in perf?

The memory perf will just be evident in the result

```
In [1]: >>> idx1 = pd.RangeIndex(0, 100_000)
   ...: >>> idx2 = pd.RangeIndex(0, 50_000)

# main
In [2]: idx1.union(idx2, sort=False).memory_usage()
Out[2]: 800000

# PR
In [2]: idx1.union(idx2, sort=False).memory_usage()
Out[2]: 128
```

**Comment by jbrockmendel** [comment]
> idx1.union

no objection to this PR, but ideally this case would be caught in _union to prevent ever materializing the ndarray

**Comment by mroeschke** [comment]
> > idx1.union
> 
> no objection to this PR, but ideally this case would be caught in _union to prevent ever materializing the ndarray

Added logic to do this in [63e7894](https://github.com/pandas-dev/pandas/pull/47557/commits/63e7894aadfc5db8462c8c4c67d960a9ed2c1731), but I think the change in `_shallow_copy` is still needed because concat goes though there

**Review by jreback — APPROVED** [review]
and asv's still hold here (after latest changes)?

**Comment by mroeschke** [comment]
> and asv's still hold here (after latest changes)?

```
% asv continuous -f 1.1 upstream/main HEAD -b index_object.Range
· Creating environments.........................
· Discovering benchmarks.
·· Uninstalling from conda-py3.8-Cython0.29.30-jinja2-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pyarrow-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt.
·· Building 498b94f7 <enh/range_shallowcopy> for conda-py3.8-Cython0.29.30-jinja2-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pyarrow-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...........................................................................
·· Installing 498b94f7 <enh/range_shallowcopy> into conda-py3.8-Cython0.29.30-jinja2-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pyarrow-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt..
· Running 20 total benchmarks (2 commits * 1 environments * 10 benchmarks)
[  0.00%] · For pandas commit 8e6ca289 <main> (round 1/2):
[  0.00%] ·· Building for conda-py3.8-Cython0.29.30-jinja2-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pyarrow-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt.........................................................................
[  0.00%] ·· Benchmarking conda-py3.8-Cython0.29.30-jinja2-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pyarrow-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[  2.50%] ··· Running (index_object.Range.time_get_loc_dec--)..........
[ 25.00%] · For pandas commit 498b94f7 <enh/range_shallowcopy> (round 1/2):
[ 25.00%] ·· Building for conda-py3.8-Cython0.29.30-jinja2-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pyarrow-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt..
[ 25.00%] ·· Benchmarking conda-py3.8-Cython0.29.30-jinja2-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pyarrow-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 27.50%] ··· Running (index_object.Range.time_get_loc_dec--)..........
[ 50.00%] · For pandas commit 498b94f7 <enh/range_shallowcopy> (round 2/2):
[ 50.00%] ·· Benchmarking conda-py3.8-Cython0.29.30-jinja2-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pyarrow-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 52.50%] ··· index_object.Range.time_get_loc_dec                                                                                                      554±100ns
[ 55.00%] ··· index_object.Range.time_get_loc_inc                                                                                                       509±50ns
[ 57.50%] ··· index_object.Range.time_iter_dec                                                                                                          14.4±2ms
[ 60.00%] ··· index_object.Range.time_iter_inc                                                                                                          15.9±1ms
[ 62.50%] ··· index_object.Range.time_max                                                                                                            1.20±0.09μs
[ 65.00%] ··· index_object.Range.time_max_trivial                                                                                                     1.10±0.2μs
[ 67.50%] ··· index_object.Range.time_min                                                                                                             1.24±0.1μs
[ 70.00%] ··· index_object.Range.time_min_trivial                                                                                                     1.48±0.3μs
[ 72.50%] ··· index_object.Range.time_sort_values_asc                                                                                                 2.73±0.2μs
[ 75.00%] ··· index_object.Range.time_sort_values_des                                                                                                   4.48±2μs
[ 75.00%] · For pandas commit 8e6ca289 <main> (round 2/2):
[ 75.00%] ·· Building for conda-py3.8-Cython0.29.30-jinja2-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pyarrow-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
[ 75.00%] ·· Benchmarking conda-py3.8-Cython0.29.30-jinja2-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pyarrow-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 77.50%] ··· index_object.Range.time_get_loc_dec                                                                                                       550±60ns
[ 80.00%] ··· index_object.Range.time_get_loc_inc                                                                                                      506±100ns
[ 82.50%] ··· index_object.Range.time_iter_dec                                                                                                          12.8±1ms
[ 85.00%] ··· index_object.Range.time_iter_inc                                                                                                          13.9±3ms
[ 87.50%] ··· index_object.Range.time_max                                                                                                             1.34±0.2μs
[ 90.00%] ··· index_object.Range.time_max_trivial                                                                                                    1.13±0.06μs
[ 92.50%] ··· index_object.Range.time_min                                                                                                             1.19±0.3μs
[ 95.00%] ··· index_object.Range.time_min_trivial                                                                                                    1.09±0.08μs
[ 97.50%] ··· index_object.Range.time_sort_values_asc                                                                                                 3.03±0.8μs
[100.00%] ··· index_object.Range.time_sort_values_des                                                                                                 5.22±0.5μs

BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```

**Comment by mroeschke** [comment]
Thanks!

**Pull request merged**